safehttp/form_test.go

```diff
@@ -0,0 +1,1051 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package safehttp
+
+import (
+	"errors"
+	"fmt"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/safehtml"
+	"github.com/google/safehtml/template"
+	"math"
+	"net/http"
+	"net/http/httptest"
+	"reflect"
+	"strconv"
+	"strings"
+	"testing"
+)
+
+const (
+	status200OK     = "200 OK"
+	status400BadReq = "400 Bad Request"
+)
+
+type dispatcher struct{}
+
+func (d *dispatcher) Write(rw http.ResponseWriter, resp Response) error {
+	switch x := resp.(type) {
+	case safehtml.HTML:
+		_, err := rw.Write([]byte(x.String()))
+		return err
+	default:
+		panic("not a safe response type")
+	}
+}
+
+func (d *dispatcher) ExecuteTemplate(rw http.ResponseWriter, t Template, data interface{}) error {
+	switch x := t.(type) {
+	case *template.Template:
+		return x.Execute(rw, data)
+	default:
+		panic("not a safe response type")
+	}
+}
+
+// A helper function that returns a parsed Form or an empty Form and any errors
+// that occured during parsing
+func getParsedForm(r *IncomingRequest) (*Form, error) {
```

Comment:
This helper should not be used. It should be clear from the test what it does, whether it asks for a PostForm or a MultipartForm.

Introducing extra logic to tests makes them complicated and error prone.

Reply:
Is it a good idea to use type assertion here in order to see whether I'm getting a Form or a MultipartForm?